File: bindings/dart/lib/src/tftp.dart
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// TFTP protocol types for proven-servers.

/// Opcode matching the Idris2 ABI tags.
enum Opcode {
  rrq(0),
  wrq(1),
  data(2),
  ack(3),
  error(4);

  const Opcode(this.tag);
  final int tag;

  static Opcode? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// TransferMode matching the Idris2 ABI tags.
enum TransferMode {
  netAscii(0),
  octet(1),
  mail(2);

  const TransferMode(this.tag);
  final int tag;

  static TransferMode? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// TftpError matching the Idris2 ABI tags.
enum TftpError {
  notDefined(0),
  fileNotFound(1),
  accessViolation(2),
  diskFull(3),
  illegalOperation(4),
  unknownTid(5),
  fileExists(6),
  noSuchUser(7);

  const TftpError(this.tag);
  final int tag;

  static TftpError? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// TransferState matching the Idris2 ABI tags.
enum TransferState {
  idle(0),
  reading(1),
  writing(2),
  inError(3),
  complete(4);

  const TransferState(this.tag);
  final int tag;

  static TransferState? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

File: bindings/javascript/src/tftp.js
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// TFTP protocol types for proven-servers.

/** Opcode matching the Idris2 ABI tags. */
export const Opcode = Object.freeze({
  RRQ: 0,
  WRQ: 1,
  DATA: 2,
  ACK: 3,
  ERROR: 4,
});

/** TransferMode matching the Idris2 ABI tags. */
export const TransferMode = Object.freeze({
  NET_ASCII: 0,
  OCTET: 1,
  MAIL: 2,
});

/** TftpError matching the Idris2 ABI tags. */
export const TftpError = Object.freeze({
  NOT_DEFINED: 0,
  FILE_NOT_FOUND: 1,
  ACCESS_VIOLATION: 2,
  DISK_FULL: 3,
  ILLEGAL_OPERATION: 4,
  UNKNOWN_TID: 5,
  FILE_EXISTS: 6,
  NO_SUCH_USER: 7,
});

/** TransferState matching the Idris2 ABI tags. */
export const TransferState = Object.freeze({
  IDLE: 0,
  READING: 1,
  WRITING: 2,
  IN_ERROR: 3,
  COMPLETE: 4,
});
